src/fsharp/FSharp.Compiler.LanguageServer/State.fs

```diff
@@ -2,11 +2,172 @@
+
+                let getArgs (tfm: string) =
+                    execArgReporter tfm
+                    |> linesWithPrefixClean detectedCommandLineArgSentinel
+
+                let tfmAndArgs =
+                    targetFrameworks
+                    |> Array.map (fun tfm -> tfm, getArgs tfm)
+
+                let separateArgs (args: string[]) =
+                    args
+                    |> Array.partition (fun a -> a.StartsWith("-"))
+                    |> (fun (options, files) ->
+                        let normalizedFiles = files |> Array.map normalizePath
+                        options, normalizedFiles)
+
+                // TODO: for now we're only concerned with the first TFM
+                let _tfm, args = Array.head tfmAndArgs
+
+                let otherOptions, sourceFiles = separateArgs args
+
+                let projectOptions: FSharpProjectOptions =
+                    { ProjectFileName = projectPath
+                      ProjectId = None
+                      SourceFiles = sourceFiles
+                      OtherOptions = otherOptions
+                      ReferencedProjects = [||] // TODO: populate from @(ProjectReference)
+                      IsIncompleteTypeCheckEnvironment = false
+                      UseScriptResolutionRules = false
+                      LoadTime = DateTime.Now
+                      UnresolvedReferences = None
+                      OriginalLoadReferences = []
+                      ExtraProjectInfo = None
+                      Stamp = None }
+                projectOptions
+            let topLevelProjects = Directory.GetFiles(rootDir, "*.fsproj")
+            let watchableProjectPaths =
+                match topLevelProjects with
+                | [||] ->
+                    match Directory.GetFiles(rootDir, "*.sln") with
+                    // TODO: what to do with multiple .sln or a combo of .sln/.fsproj?
+                    | [| singleSolution |] ->
+                        let sln = SolutionFile.Parse(singleSolution)
+                        let projects =
+                            sln.ProjectsInOrder
+                            |> Seq.map (fun p -> p.AbsolutePath)
+                            |> Seq.filter (fun p -> Path.GetExtension(p) = ".fsproj")
+                            |> Seq.toArray
+                        projects
+                    | _ -> [||]
+                | _ -> topLevelProjects
+            let watchableProjectOptions =
+                watchableProjectPaths
+                |> Array.map getProjectOptions
+
+            // associate source files with project options
+            // TODO: watch for changes to .fsproj
+            // TODO: watch for changes to .deps.json
+            watchableProjectOptions
+            |> Seq.iter (fun projectOptions ->
+                projectOptions.SourceFiles
+                |> Seq.iter (fun sourceFile -> sourceFileToProjectMap.AddOrUpdate(sourceFile, projectOptions, fun _ _ -> projectOptions) |> ignore ))
```

Comment:
Why not `for ... in ... do for ... in ... do ...`, as several lines below?

Reply:
Good catch, updated.